File: RiscVEmulator-RV32I/RiscVEmulator.Tests/StringTest.cs
using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace RiscVEmulator.Tests;

[TestClass]
public class StringTest : EmulatorTestBase
{
    [TestMethod]
    public void String_AllFunctions()
    {
        string output = RunProgramWithLibc("string_test");

        Assert.IsTrue(output.Contains("strlen_empty:0"));
        Assert.IsTrue(output.Contains("strlen_hello:5"));
        Assert.IsTrue(output.Contains("strcmp_eq:1"));
        Assert.IsTrue(output.Contains("strcmp_lt:1"));
        Assert.IsTrue(output.Contains("strcmp_gt:1"));
        Assert.IsTrue(output.Contains("strncmp_eq:1"));
        Assert.IsTrue(output.Contains("strncmp_ne:1"));
        Assert.IsTrue(output.Contains("strcpy:hello"));
        Assert.IsTrue(output.Contains("strncpy:hi"));
        Assert.IsTrue(output.Contains("strncpy_pad:1"));
        Assert.IsTrue(output.Contains("strcat:hello world"));
        Assert.IsTrue(output.Contains("strncat:foobar"));
        Assert.IsTrue(output.Contains("strchr:l"));
        Assert.IsTrue(output.Contains("strchr_null:1"));
        Assert.IsTrue(output.Contains("strrchr:l"));
        Assert.IsTrue(output.Contains("strstr:world"));
        Assert.IsTrue(output.Contains("strstr_null:1"));
        Assert.IsTrue(output.Contains("memcmp_eq:1"));
        Assert.IsTrue(output.Contains("memcmp_ne:1"));
        Assert.IsTrue(output.Contains("memchr:l"));
        Assert.IsTrue(output.Contains("memchr_null:1"));
        Assert.IsTrue(output.Contains("memmove:ababcdef"));
        Assert.IsTrue(output.Contains("PASS"));
    }
}

File: RiscVEmulator-RV32IMA_Zicsr/RiscVEmulator.Tests/Programs/string_test.c
#include "libc.h"

void _start(void) {
    // strlen
    printf("strlen_empty:%d\n", (int)strlen(""));
    printf("strlen_hello:%d\n", (int)strlen("hello"));

    // strcmp
    printf("strcmp_eq:%d\n", strcmp("abc", "abc") == 0);
    printf("strcmp_lt:%d\n", strcmp("abc", "abd") < 0);
    printf("strcmp_gt:%d\n", strcmp("abd", "abc") > 0);

    // strncmp
    printf("strncmp_eq:%d\n", strncmp("abcdef", "abcxyz", 3) == 0);
    printf("strncmp_ne:%d\n", strncmp("abcdef", "abcxyz", 4) != 0);

    // strcpy
    char buf[32];
    strcpy(buf, "hello");
    printf("strcpy:%s\n", buf);

    // strncpy
    char buf2[8];
    strncpy(buf2, "hi", 8);
    printf("strncpy:%s\n", buf2);
    printf("strncpy_pad:%d\n", buf2[3] == '\0');  // should be zero-padded

    // strcat
    char cat[32];
    strcpy(cat, "hello");
    strcat(cat, " world");
    printf("strcat:%s\n", cat);

    // strncat
    char ncat[32];
    strcpy(ncat, "foo");
    strncat(ncat, "barbaz", 3);
    printf("strncat:%s\n", ncat);

    // strchr
    printf("strchr:%c\n", *strchr("hello", 'l'));
    printf("strchr_null:%d\n", strchr("hello", 'z') == NULL);

    // strrchr
    printf("strrchr:%c\n", *strrchr("hello", 'l'));

    // strstr
    printf("strstr:%s\n", strstr("hello world", "world"));
    printf("strstr_null:%d\n", strstr("hello", "xyz") == NULL);

    // memcmp
    printf("memcmp_eq:%d\n", memcmp("abc", "abc", 3) == 0);
    printf("memcmp_ne:%d\n", memcmp("abc", "abd", 3) < 0);

    // memchr
    printf("memchr:%c\n", *(char *)memchr("hello", 'l', 5));
    printf("memchr_null:%d\n", memchr("hello", 'z', 5) == NULL);

    // memmove (overlapping)
    char mv[16] = "abcdefgh";
    memmove(mv + 2, mv, 6);  // "ababcdef"
    mv[8] = '\0';
    printf("memmove:%s\n", mv);

    printf("PASS\n");
    exit(0);
}
